docs/RFCs/19042021_Cert_Management_In_CRDB_Helm_Chart.md

```diff
@@ -0,0 +1,243 @@
+  the new CA certificate followed by the old CA certificate.
+* This new combined CA certificate will be updated in the CA secret and all the Node and Client secret.
+  `TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
+* In addition, Client secret and Node secret will be patched with annotation `needsRegeneration: True`, which specifies that their certificates
+  need to be regenerated in the next cronjob run. This is done in accordance with the suggestion in CockroachDB [doc](https://www.cockroachlabs.com/docs/v20.2/rotate-certificates.html#why-rotate-ca-certificates-in-advance)
+ 
+#### Check-node-cert-for-regeneration
+* Follow checks from [Check cert for regeneration](#check-cert-for-regeneration)
+* In addition, check if annotation `needsRegeneration: True` exists, if yes return True (This will be case when CA certificate has rotated, but Client and node certs are still to be recreated)
+ 
+#### Check-cert-for-regeneration
+* Check if the certificate secret has annotation `managed-by: cockroachdb`, if not return False
+* Check if the certificate is expiring in the next 2 months using the values from annotations `creationTime` and `duration` on secret, if expiring return True, else False
+ 
+#### Validate-Secret-Annotations
+* Check if annotation for `resourceVersion` and `creationTime` exists, if not return False
+* Check if resourceVersion of the secret matches with the value of `resourceVersion` annotation, if not means secret is changed, return False
+* Else, return True
+ 
+### Certificate Generation cases during Helm upgrade:
+ 
+In case of Helm upgrade:
+ 
+* User has given CA and changes contents of the CA secret:
+  * Check if the current value `resourceVersion` or hash matches with the annotation value. if annotation does not match, so this is a new CA scenario.
+* User has given CA and changes the secret name:
+  * Annotation will not be found, so this is a new CA scenario.
+* User had not given CA previously, but now has given the CA:
+  * Annotation will not be found, so this is a new CA scenario
+* If the user changes the duration of CA:
+  * Identify and compare using existing annotation on CA secret and current value, this will be a case for certificate rotation. This will only be the case when the CA is managed by CockroachDB.
+* User changes the duration of all certificates:
+  * Compare old and new CA duration from CA secret annotation values and current value. This will be a case for certificate rotation.
+  * Rotate CA certificate and add an annotation on CA secret with the date of rotation.
+  * Add annotations on Node and Client certs specifying the new expected duration and `to-be-rotated: true`.
+  * These secret certificates will be renewed in the next cron cycle and `to-be-rotated:true` and expected duration annotations will be removed.
```

Comment:
This annotation will be generated by our code and is no user input. We have a separate helm input that decides if certs are to be rotated or not. Cron jobs will only be created in case cert-rotation is set to True via helm input